### PR TITLE
Allow upload of TimeseriesDataset with NaN values

### PR DESCRIPTION
**Pull request by LeenaPajo**
# Description

Please include a summary of the change. Please also include relevant motivation and context. This can be the same as your commit message if it contains those points.

Fixes # (issue).  *(remove if not applicable)*

# Additional information

Please add additional information for the reviewer here. 

Examples: integration test setup including execution environment and steps for reproduction, images of plots, etc.

# Checklist

Mark "not applicable" if item on the list does not apply to this pull request.

- [x] I have created/adapted unit tests for new code
  - [ ] not applicable 
- [ ] I have added new dependencies as described at the [Contributing](https://sap.github.io/project-sailor/contributing.html#requirements-management) page
  - [x] not applicable 
- [ ] I have made corresponding changes to the documentation (incl. tutorial, etc.)
  - [ ] not applicable 
- [ ] I have provided release notes (in description or as comment) if this PR is a new feature OR a change worth mentioning for next release
  - [x] not applicable 
- [ ] I have obtained API approvals if a new SAP API or endpoint is used
  - [ ] not applicable

### CODE REVIEW

File: tests/test_sailor/test_sap_iot/test_write.py [hunk 1]
```diff
@@ -88,6 +89,28 @@ def test_each_equipment_one_request(mock_request, mock_upload_url, make_indicato
...
+                                       pstid=['indicator_group_A'],)
+    equipment_set = make_equipment_set(equipmentId=['equipment_A'])
+    dataset = make_dataset(indicator_set, equipment_set, 2)
+    dataset._df.iloc[0, 2] = np.nan
+    upload_indicator_data(dataset)
+
+    payloads = [args[-1]['json'] for args in mock_request.call_args_list]
+    indicator_values = []
+    for payload in payloads:
+        for values in payload.values():
```
Comment (atrejut) [inline tests/test_sailor/test_sap_iot/test_write.py:103]: why the loop? You're looking also at `payload['Tags']` this way, when you could just do `payload['Values']` instead?

File: tests/test_sailor/test_sap_iot/test_write.py [hunk 2]
```diff
@@ -88,6 +89,28 @@ def test_each_equipment_one_request(mock_request, mock_upload_url, make_indicato
     assert urls == {request_base + equipment.id for equipment in equipment_set}
 
 
+def test_nan_dataset_written(mock_request, make_indicator_set, make_equipment_set):
+    indicator_set = make_indicator_set(propertyId=['indicator_id_A', 'indicator_id_B'],
+                                       pstid=['indicator_group_A'],)
```
Comment (atrejut) [inline tests/test_sailor/test_sap_iot/test_write.py:94]: why specify the `pstid` explicitly? `propertyId` should be enough, no?

File: tests/test_sailor/test_sap_iot/test_write.py [hunk 3]
```diff
@@ -88,6 +89,28 @@ def test_each_equipment_one_request(mock_request, mock_upload_url, make_indicato
...
+    dataset._df.iloc[0, 2] = np.nan
+    upload_indicator_data(dataset)
+
+    payloads = [args[-1]['json'] for args in mock_request.call_args_list]
+    indicator_values = []
+    for payload in payloads:
+        for values in payload.values():
+            for x in values:
+                if 'I_indicator_id_A' in x:
+                    indicator_values.append(x['I_indicator_id_A'])
```
Comment (atrejut) [inline tests/test_sailor/test_sap_iot/test_write.py:106]: this seems fragile in terms of ordering. I'd rather have `indicator_values_A` and `indicator_values_B` or some such.

File: tests/test_sailor/test_sap_iot/test_write.py [hunk 4]
```diff
@@ -88,6 +89,28 @@ def test_each_equipment_one_request(mock_request, mock_upload_url, make_indicato
     assert urls == {request_base + equipment.id for equipment in equipment_set}
 
 
+def test_nan_dataset_written(mock_request, make_indicator_set, make_equipment_set):
+    indicator_set = make_indicator_set(propertyId=['indicator_id_A', 'indicator_id_B'],
+                                       pstid=['indicator_group_A'],)
+    equipment_set = make_equipment_set(equipmentId=['equipment_A'])
+    dataset = make_dataset(indicator_set, equipment_set, 2)
+    dataset._df.iloc[0, 2] = np.nan
```
Comment (atrejut) [inline tests/test_sailor/test_sap_iot/test_write.py:97]: this is fragile, depends on the order of indicators in `_df`...

**Review by atrejut — COMMENTED** [review]
pretty sure your test only works by accident. You're not actually testing for a `None` in `indicator A`, you're just testing for a `None` in the first indicator you happen to come across.
However, since you're also not setting `np.NaN` for `indicator A`, but rather just for the first indicator in the dataframe it actually works out.

Still I'd much prefer to not rely on the order, but do things by row/column index.

here's a proposal that also changes the `asserts` in a way I find more easy to understand, but feel free to take or ignore that part:

```
    indicator_set = make_indicator_set(propertyId=['indicator_id_A', 'indicator_id_B'])
    equipment_set = make_equipment_set(equipmentId=['equipment_A'])
    dataset = make_dataset(indicator_set, equipment_set, 2)

    none_indicator = indicator_set[0]
    valid_indicator = indicator_set[1]
    none_timestamp = _timestamp_to_isoformat(dataset._df.loc[0, 'timestamp'], with_zulu=True)
    dataset._df.loc[0, none_indicator._unique_id] = np.nan

    upload_indicator_data(dataset)

    payloads = [args[-1]['json'] for args in mock_request.call_args_list]
    for payload in payloads:
        for values_at_timestamp in payload['Values']:
            assert isinstance(values_at_timestamp[valid_indicator._liot_id], float)
            
            if values_at_timestamp['_time'] == none_timestamp:
                assert values_at_timestamp[none_indicator._liot_id] is None
            else:
                assert isinstance(values_at_timestamp[none_indicator._liot_id], float)
```

**Review by cloudyday — APPROVED** [review]
LGTM

**Pull request merged**